From: Egor Ignatov <egori@altlinux.org>
Date: Thu, 23 Jan 2025 20:44:15 +0300
Subject: fs/xfs: Propagate incorrect inode error from grub_xfs_read_inode
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

The incorrect inode error from grub_xfs_read_inode did not propagate because
grub_print_error() resetted grub_errno, and grub_xfs_iterate_dir() did not
handle it at all.

Signed-off-by: Egor Ignatov <egori@altlinux.org>
FG: cherry-picked from grub-devel 20250123174415.1251915-2-egori@altlinux.org
    with adapted context
Signed-off-by: Fabian Grünbichler <f.gruenbichler@proxmox.com>
---
 grub-core/fs/xfs.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/grub-core/fs/xfs.c b/grub-core/fs/xfs.c
index e80a7b0..98fd53b 100644
--- a/grub-core/fs/xfs.c
+++ b/grub-core/fs/xfs.c
@@ -777,7 +777,6 @@ static int iterate_dir_call_hook (grub_uint64_t ino, const char *filename,
   fdiro = grub_malloc (sz);
   if (!fdiro)
     {
-      grub_print_error ();
       return 0;
     }
 
@@ -789,7 +788,6 @@ static int iterate_dir_call_hook (grub_uint64_t ino, const char *filename,
   err = grub_xfs_read_inode (ctx->diro->data, ino, &fdiro->inode);
   if (err)
     {
-      grub_print_error ();
       return 0;
     }
 
@@ -828,9 +826,13 @@ grub_xfs_iterate_dir (grub_fshelp_node_t dir,
 	/* Synthesize the direntries for `.' and `..'.  */
 	if (iterate_dir_call_hook (diro->ino, ".", &ctx))
 	  return 1;
+	else if (grub_errno)
+	  return 0;
 
 	if (iterate_dir_call_hook (parent, "..", &ctx))
 	  return 1;
+	else if (grub_errno)
+	  return 0;
 
 	for (i = 0; i < head->count; i++)
 	  {
@@ -870,6 +872,9 @@ grub_xfs_iterate_dir (grub_fshelp_node_t dir,
 	      }
 	    de->name[de->len] = c;
 
+	    if (grub_errno)
+	      return 0;
+
 	    de = grub_xfs_inline_next_de(dir->data, head, de);
 
 	    if ((grub_uint8_t *) de >= (grub_uint8_t *) dir + grub_xfs_fshelp_size (dir->data))
@@ -955,6 +960,11 @@ grub_xfs_iterate_dir (grub_fshelp_node_t dir,
 		    grub_free (dirblock);
 		    return 1;
 		  }
+		else if (grub_errno)
+		  {
+		    grub_free (dirblock);
+		    return 0;
+		  }
 
 		/* Check if last direntry in this block is
 		   reached.  */
